### ISSUE TITLE
Time "Top Up" Limits

### ISSUE THREAD

**Issue by 0x4007** [ubiquity-os-marketplace/daemon-disqualifier#51]
A quick idea that can use improvement:

Assignee can restart the disqualification timer by continuing to try. There's many clear examples of the task not being the right fit for the assignee and they try i.e. for months on projects that should be doable in a day etc. 

We could consider doing limits on time limit "top ups." So for example:

Priority 1 - 5 top ups
Priority 2 - 4 top ups 
Priority 3 - 3 top ups 
Priority 4 - 2 top ups 
Priority 5 - 1 top up. 

I suspect that priority 5 tasks will border on eliminating most even capable assignees but once the task is complete even if just slightly after deadline we can always retroactively assign credit. 

---

- We should be able tot toggle this behavior in the config. 
- if it is enabled the message should clearly explain the amount of remaining top ups. 

I foresee a bit of complexity around the logic related to dealing with the "time limits" for follow up/disqualify and the top up counter. 
- If the limits are 7/14 as they are now, then the first top up should occur after 14 days on a priority 1 normal project. 
- priority 5 top up would be after `14 / 5 = 2.8` days.

**Comment by gentlementlegen** [comment]
@0x4007 Wouldn't it be simpler to use a "quota" or "credits" for reminders?

Based on what you describe:
- priority 1 gets 5 credits
- priority 2 gets 4 credits
- etc...

So basically you would be allowed to get `n` reminders total for a given task, each of the reminders using 1 credit?

Maybe I didn't understand properly, but given the specification, consider the following example:

- a user has 7 days to fulfill the task
- the user takes 6 days to finish it, and changes it to "ready for review"
- no review happens within 3.5 days, a reminder should be sent, but instead the user gets disqualified right away?

Is this correct?

**Comment by 0x4007** [comment]
The concept of "credits" seem to imply that storage is required. My strategy is just a calculation based on elapsed time. 

You might be able to parse the entire timeline every time to see if there are exceptions to the timing though, like for slow reviews and such.

**Comment by gentlementlegen** [comment]
I think your idea makes more sense because with mine, a `Priority 1` with a `1 Week` deadline would get 5 reminders thus a month of extra time at most, which is too much for a one week tasks obviously. 

> Check the assignment timestamp and use that as a point of reference to calculate the current top up counter/state. i.e. if the task was assigned 35 days ago, and the clock speed is 14 days, then 35/14=2.5 meaning we are still at top up number 2.

However wouldn't this mean that it would also be the case with your spec? I know the deadlines are not meant to be strictly exact, but the gap between a week and a month seems too big.

Maybe we could make it as simple as one reminder allowed only? Would that be too harsh? If we handle slow reviews I think it would help us keeping deadline tighter.

**Comment by 0x4007** [comment]
Of course in an ideal world everybody processes their tasks with utmost priority and speed. But the reality is that even our core team members spend several weeks on, for example, a one week task. 

After monitoring these behaviors for some years I confidently came up with this design for the specification.

**Comment by gentlementlegen** [comment]
Yes I understand the problem you are trying to solve. The one I am thinking of is when users take a 1h task, and they keep pushing for many days, so no reminder or blame is sent although that task should have been finished quite quickly. I've seen 15 minutes tasks take a week.

Let's try implementing your idea and see how it goes, it just feels like more pressure is put on high priority tasks which are currently mostly handled by the core team, when I think what you're targeting are low level priority tasks done by outside contributors.

**Comment by 0x4007** [comment]
I would like to try the current proposed strategy first but we can potentially have a static amount of top ups too with an alternative configuration (for example, 5 for every priority level) which would make things easier for the higher priority tasks.

**Comment by 0x4007** [comment]
/ask based on the specification please provide some typescript code for the developer to start this implementation with

**Comment by gentlementlegen** [comment]
I am convinced that the top-up calculation can be made much simpler, because in the end it is just a multiplier for the deadline, no need to calculate all the in-between steps, more likely just the "final final" deadline,

**Comment by 0x4007** [comment]
> no need to calculate all the in-between steps

What about the reminders?

**Comment by ManyRios** [comment]
in this example this issue... the pty is 4 but it needs to be done in less than 1 day.... 

I think we can take the ```priorityTopUps ```  and re-use some functions that are already here in the repo.

**Comment by gentlementlegen** [comment]
@0x4007 Currently, if a reminder is sent, we do not send anymore reminders until the final deadline, or if user activity has happened in between. I guess this should be change to remind many times due to the top-up system?

**Comment by 0x4007** [comment]
Yeah I think each top up should incur a reminder of sorts

**Comment by gentlementlegen** [comment]
Currently I just mention it in the reminder "Top ups remaining: 3" but that can be tweaked.

**Comment by 0x4007** [comment]
I think there is value in saying both numbers to take advantage of the psychological principle of loss aversion.

**Comment by 0x4007** [comment]
Your formatting score needs better rounding. Can you fix that? I've seen it a few times before. @gentlementlegen

**Comment by gentlementlegen** [comment]
It is current set to 3 decimals. Want it to two?
